Adopt new ServiceExtensions scroll view delegate API

whsieh opened the pull request:
#### 75f37688a521db96bf5943381c323eb9acb1f0c6
<pre>
Adopt new ServiceExtensions scroll view delegate API
<a href="https://bugs.webkit.org/show_bug.cgi?id=267214">https://bugs.webkit.org/show_bug.cgi?id=267214</a>
<a href="https://rdar.apple.com/120597428">rdar://120597428</a>

Reviewed by Tim Horton.

Adopt a couple of scroll view delegate APIs in `WKScrollingNodeScrollViewDelegate` and `WKWebView`:

```
-scrollView:handleScrollUpdate:completion:
-parentScrollViewForScrollView:
```

...instead of equivalent SPI subclassing hooks. To do this, we replace various UIKit SPI symbols
with equivalent declarations:

```
UIScrollEvent            →   WKSEScrollViewScrollUpdate
UIScrollPhase            →   WKSEScrollViewScrollUpdatePhase
UIScrollPhaseBegan       →   WKSEScrollViewScrollUpdatePhaseBegan
UIScrollPhaseChanged     →   WKSEScrollViewScrollUpdatePhaseChanged
UIScrollPhaseEnded       →   WKSEScrollViewScrollUpdatePhaseEnded
UIScrollPhaseCancelled   →   WKSEScrollViewScrollUpdatePhaseCancelled
```

See below for more details.

* Source/WebKit/Shared/ios/WebIOSEventFactory.h:
* Source/WebKit/Shared/ios/WebIOSEventFactory.mm:

Add a missing `WebKit` namespace around `WebIOSEventFactory`, and deploy the new WKSE-prefixed
scroll view symbols (see above).

(WebKit::WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags):
(WebKit::WebIOSEventFactory::toUIKeyModifierFlags):
(WebKit::modifiersForEvent):
(WebKit::WebIOSEventFactory::createWebKeyboardEvent):
(WebKit::WebIOSEventFactory::createWebMouseEvent):
(WebKit::toWebPhase):

Note: since `UIScrollPhaseNone` and `UIScrollPhaseMayBegin` don&apos;t have WKSE-prefixed equivalents,
this codepath is limited to the non-ServiceExtensions case.

(WebKit::WebIOSEventFactory::translationInView):
(WebKit::WebIOSEventFactory::createWebWheelEvent):
(WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags): Deleted.
(WebIOSEventFactory::toUIKeyModifierFlags): Deleted.
(modifiersForEvent): Deleted.
(WebIOSEventFactory::createWebKeyboardEvent): Deleted.
(WebIOSEventFactory::createWebMouseEvent): Deleted.
(toWebPhase): Deleted.
(WebIOSEventFactory::createWebWheelEvent): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction modifierFlags]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _scrollView:asynchronouslyHandleScrollEvent:completion:]):
(-[WKWebView scrollView:handleScrollUpdate:completion:]):
(-[WKWebView scrollView:handleScrollEvent:completion:]): Deleted.

Add an implementation of `-scrollView:handleScrollUpdate:completion:`, and have the legacy SPI
delegate method call into this new method. Similarly, have `-_actingParentScrollViewForScrollView:`
call into the new `-parentScrollViewForScrollView:` method.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::mouseDidMoveOverElement):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(WebKit::findActingScrollParent):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(-[WKScrollingNodeScrollViewDelegate _actingParentScrollViewForScrollView:]):
(-[WKScrollingNodeScrollViewDelegate parentScrollViewForScrollView:]):
(-[WKScrollingNodeScrollViewDelegate _scrollView:asynchronouslyHandleScrollEvent:completion:]):
(-[WKScrollingNodeScrollViewDelegate scrollView:handleScrollUpdate:completion:]):
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::handleAsynchronousCancelableScrollEvent):
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::findActingScrollParent):
(-[WKScrollingNodeScrollViewDelegate actingParentScrollViewForScrollView:]): Deleted.
(-[WKScrollingNodeScrollViewDelegate scrollView:handleScrollEvent:completion:]): Deleted.
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::handleAsynchronousCancelableScrollEvent):
* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView initWithFrame:]):
(-[WKBaseScrollView _subclassHandlesAsyncScrollEvent]): Deleted.
(-[WKBaseScrollView _asynchronouslyHandleScrollEvent:completion:]): Deleted.

Remove these subclass method implementations, which are no longer necessary after adopting the new
delegate methods.

(-[WKBaseScrollView _actingParentScrollView]): Deleted.
* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
(-[WKMouseInteraction createMouseEventWithType:wasCancelled:]):
* Source/WebKit/UIProcess/ios/WKSEDefinitions.h:

Add fallback declarations for the new scroll-view-related `WKSE*` symbols.

* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(legacyScrollPhase):
(createScrollUpdate):

Update these API tests to call into the new delegate methods and create `WKSEScrollViewScrollUpdate`
instead of `UIScrollEvent` if needed.

Canonical link: <a href="https://commits.webkit.org/273026@main">https://commits.webkit.org/273026@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d6ff3f6c523d1f622371390ceb8e6a7ca42dd29

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12796 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36639 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30818 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35074 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15180 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9937 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29869 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34504 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10820 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30340 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9439 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30360 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37947 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30887 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30683 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35652 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9671 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7600 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33555 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11457 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7831 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10236 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10482 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/ios/WebIOSEventFactory.mm:
```diff
@@ -145,57 +147,71 @@
 {
     switch (phase) {
+#if !SERVICE_EXTENSIONS_SCROLL_VIEW_IS_AVAILABLE
     case UIScrollPhaseNone:
```
Matching `WKSEScrollViewScrollUpdatePhase` values against `UIScrollPhase` literals?

aprotyas replied:
Okay, I see what happened here. I thought the compiler would complain but they're all just integral values. 🤷

whsieh replied:
Indeed, `UIScrollPhaseNone` does not have a `WKSE`-prefixed equivalent. In the case where `SERVICE_EXTENSIONS_SCROLL_VIEW_IS_AVAILABLE` is off, `WKSEScrollViewScrollUpdatePhase` is typedef'd to be `UIScrollPhase`, so this is identical to shipping code.

aprotyas reviewed (commented):
LGTM

whsieh committed:
Adopt new ServiceExtensions scroll view delegate API
https://bugs.webkit.org/show_bug.cgi?id=267214
rdar://120597428

Reviewed by Tim Horton.

Adopt a couple of scroll view delegate APIs in `WKScrollingNodeScrollViewDelegate` and `WKWebView`:

```
-scrollView:handleScrollUpdate:completion:
-parentScrollViewForScrollView:
```

...instead of equivalent SPI subclassing hooks. To do this, we replace various UIKit SPI symbols
with equivalent declarations:

```
UIScrollEvent            →   WKSEScrollViewScrollUpdate
UIScrollPhase            →   WKSEScrollViewScrollUpdatePhase
UIScrollPhaseBegan       →   WKSEScrollViewScrollUpdatePhaseBegan
UIScrollPhaseChanged     →   WKSEScrollViewScrollUpdatePhaseChanged
UIScrollPhaseEnded       →   WKSEScrollViewScrollUpdatePhaseEnded
UIScrollPhaseCancelled   →   WKSEScrollViewScrollUpdatePhaseCancelled
```

See below for more details.

* Source/WebKit/Shared/ios/WebIOSEventFactory.h:
* Source/WebKit/Shared/ios/WebIOSEventFactory.mm:

Add a missing `WebKit` namespace around `WebIOSEventFactory`, and deploy the new WKSE-prefixed
scroll view symbols (see above).

(WebKit::WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags):
(WebKit::WebIOSEventFactory::toUIKeyModifierFlags):
(WebKit::modifiersForEvent):
(WebKit::WebIOSEventFactory::createWebKeyboardEvent):
(WebKit::WebIOSEventFactory::createWebMouseEvent):
(WebKit::toWebPhase):

Note: since `UIScrollPhaseNone` and `UIScrollPhaseMayBegin` don't have WKSE-prefixed equivalents,
this codepath is limited to the non-ServiceExtensions case.

(WebKit::WebIOSEventFactory::translationInView):
(WebKit::WebIOSEventFactory::createWebWheelEvent):
(WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags): Deleted.
(WebIOSEventFactory::toUIKeyModifierFlags): Deleted.
(modifiersForEvent): Deleted.
(WebIOSEventFactory::createWebKeyboardEvent): Deleted.
(WebIOSEventFactory::createWebMouseEvent): Deleted.
(toWebPhase): Deleted.
(WebIOSEventFactory::createWebWheelEvent): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction modifierFlags]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _scrollView:asynchronouslyHandleScrollEvent:completion:]):
(-[WKWebView scrollView:handleScrollUpdate:completion:]):
(-[WKWebView scrollView:handleScrollEvent:completion:]): Deleted.

Add an implementation of `-scrollView:handleScrollUpdate:completion:`, and have the legacy SPI
delegate method call into this new method. Similarly, have `-_actingParentScrollViewForScrollView:`
call into the new `-parentScrollViewForScrollView:` method.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::mouseDidMoveOverElement):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(WebKit::findActingScrollParent):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(-[WKScrollingNodeScrollViewDelegate _actingParentScrollViewForScrollView:]):
(-[WKScrollingNodeScrollViewDelegate parentScrollViewForScrollView:]):
(-[WKScrollingNodeScrollViewDelegate _scrollView:asynchronouslyHandleScrollEvent:completion:]):
(-[WKScrollingNodeScrollViewDelegate scrollView:handleScrollUpdate:completion:]):
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::handleAsynchronousCancelableScrollEvent):
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::findActingScrollParent):
(-[WKScrollingNodeScrollViewDelegate actingParentScrollViewForScrollView:]): Deleted.
(-[WKScrollingNodeScrollViewDelegate scrollView:handleScrollEvent:completion:]): Deleted.
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::handleAsynchronousCancelableScrollEvent):
* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView initWithFrame:]):
(-[WKBaseScrollView _subclassHandlesAsyncScrollEvent]): Deleted.
(-[WKBaseScrollView _asynchronouslyHandleScrollEvent:completion:]): Deleted.

Remove these subclass method implementations, which are no longer necessary after adopting the new
delegate methods.

(-[WKBaseScrollView _actingParentScrollView]): Deleted.
* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
(-[WKMouseInteraction createMouseEventWithType:wasCancelled:]):
* Source/WebKit/UIProcess/ios/WKSEDefinitions.h:

Add fallback declarations for the new scroll-view-related `WKSE*` symbols.

* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(legacyScrollPhase):
(createScrollUpdate):

Update these API tests to call into the new delegate methods and create `WKSEScrollViewScrollUpdate`
instead of `UIScrollEvent` if needed.

Canonical link: https://commits.webkit.org/273026@main
